source/lac/constraint_matrix.cc

```diff
@@ -530,56 +531,85 @@ ConstraintMatrix::merge (const ConstraintMatrix &other_constraints,
                 tmp.push_back (std::pair<size_type,double>(j->first,
                                                            j->second*weight));
 
-              line->inhomogeneity += other_constraints.get_inhomogeneity(line->entries[i].first) *
-                                     weight;
+              line->inhomogeneity
+              += other_constraints.get_inhomogeneity(line->entries[i].first) *
+                 weight;
             }
         }
       // finally exchange old and newly resolved line
       line->entries.swap (tmp);
     }
 
-
-
-  // next action: append those lines at the end that we want to add
-  for (std::vector<ConstraintLine>::const_iterator
-       line=other_constraints.lines.begin();
-       line!=other_constraints.lines.end(); ++line)
-    if (is_constrained(line->line) == false)
-      lines.push_back (*line);
+  {
+    size_type new_size;
+    if (local_lines.size() != 0)
+      {
+        Assert (other_constraints.local_lines.size() !=0, ExcNotImplemented());
+        // The last stored index determines the new size of lines_cache
+        const size_type last_stored_index
+          = std::max(local_lines.nth_index_in_set(lines_cache.size()-1),
```

Comment:
Since this PR was merged I see assertions from this line when running ASPECT models, because we have the case that `local_lines.size()` is larger than 0 (so we end up inside of this if-statement), but `lines_cache.size() == 0` so this call to `nth_index_in_set` is asking for the 4294967295th index. I know too little about the internals of `ConstraintMatrix` to immediately see what is going wrong here, but maybe @masterleinad you have an idea? Otherwise I will try to think about this somewhat more.